### PR TITLE
Add `DefinedMacros` handler

### PR DESCRIPTION
**Pull request by Draggu**
**Stack**:
- #1686
- #1685
- #1684
- #1683 ⬅
- #1718

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: scarb/src/ops/proc_macro_server/methods/defined_macros.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        let mut response: DefinedMacrosResponse = proc_macros
+            .macros()
+            .into_iter()
+            .map(|e| DefinedMacrosResponse {
+                attributes: e.declared_attributes(),
+                inline_macros: e.inline_macros(),
+                derives: e.declared_derives(),
+                executables: e.executable_attributes(),
+            })
+            .sum();
```
Comment (mkaput) [inline scarb/src/ops/proc_macro_server/methods/defined_macros.rs:19]: yeah, the summing procedure should happen entirely in this file, not in the `_api` crate. especially that you're layering extra logic with that `retain` right afterwards

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/ops/proc_macro_server/methods/defined_macros.rs
Comment (mkaput) [inline scarb/src/ops/proc_macro_server/methods/defined_macros.rs:1]: missing tests

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/tests/proc_macro_server.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            ",
+        )
+        .dep("some", plugin_package)
+        .build(&project);
+
+    let mut proc_macro_server = TestProcMacroServer::new(&project);
+
+    let request = proc_macro_server.request::<DefinedMacros>(DefinedMacrosParams {});
+
+    let response = proc_macro_server.wait_for_response(request).ok();
```
Comment (mkaput) [inline scarb/tests/proc_macro_server.rs:39]: sounds like chaining these two methods will be a common pattern? perhaps you could add a shortcut for it in `TestProcMacroServer` ?

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — APPROVED** [review]

File: scarb/src/ops/proc_macro_server/json_rpc.rs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use scarb_proc_macro_server_types::methods::Method;
+
+use crate::compiler::plugin::proc_macro::ProcMacroHost;
+
+pub trait Handler: Method {
+    fn handle(proc_macros: Arc<ProcMacroHost>, params: Self::Params) -> Result<Self::Response>;
```
Comment (piotmag769) [inline scarb/src/ops/proc_macro_server/json_rpc.rs:9]:
```suggestion
    fn handle(proc_macro_host: Arc<ProcMacroHost>, params: Self::Params) -> Result<Self::Response>;
```

File: scarb/src/ops/proc_macro_server/json_rpc.rs
Comment (piotmag769) [inline scarb/src/ops/proc_macro_server/json_rpc.rs:1]: Wouldn't it be better in `scarb/src/ops/proc_macro_server/mod.rs` or `scarb/src/ops/proc_macro_server/{request_handlers/MAYBE_DIFFERENT_NAME}/mod.rs`? Logically this module should be the ancestor of handler modules. Then you can remove `pub`s etc.

File: scarb/src/ops/proc_macro_server/methods/defined_macros.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        response.derives.sort();
+        response.derives.dedup();
+
+        response.executables.sort();
+        response.executables.dedup();
+
+        response
+            .attributes
+            .retain(|attr| !response.executables.contains(attr));
```
Comment (piotmag769) [inline scarb/src/ops/proc_macro_server/methods/defined_macros.rs:46]: Should be a different method for getting attribute macros only so we don't need to do that

File: scarb/src/ops/proc_macro_server/mod.rs [hunk 5]
```diff
@@ -56,10 +68,14 @@ fn handle_requests(receiver: Receiver<RpcRequest>, sender: Sender<RpcResponse>)
...
-    #[allow(clippy::match_single_binding)]
+fn route_request(proc_macros: Arc<ProcMacroHost>, request: RpcRequest) -> Result<Value> {
     match request.method.as_str() {
-        //TODO add method handlers
+        DefinedMacros::METHOD => run_handler::<DefinedMacros>(proc_macros.clone(), request.value),
         _ => Err(anyhow!("method not found")),
     }
 }
+
+fn run_handler<M: Handler>(proc_macros: Arc<ProcMacroHost>, value: Value) -> Result<Value> {
```
Comment (piotmag769) [inline scarb/src/ops/proc_macro_server/mod.rs:78]: [nit]
```suggestion
fn handle<M: Handler>(proc_macros: Arc<ProcMacroHost>, value: Value) -> Result<Value> {
```

Comment (Draggu) [reply scarb/src/ops/proc_macro_server/mod.rs:78]: we already have `handle_requests()` and `Handler::handle()`, naming next function `handle` will make it less readable imo, now it is clear that we will execute `Handler::handle()`

**Review by piotmag769 — APPROVED** [review]

**Pull request merged**